vpr/src/place/place_constraints.cpp

```diff
@@ -82,29 +82,69 @@ PartitionRegion constrained_macro_locs(const t_pl_macro& pl_macro) {
 
                 t_pl_loc min_pl_loc(reg_rect.xmin(), reg_rect.ymin(), block_regions[i].get_sub_tile());
 
-                t_pl_loc modified_min_pl_loc = min_pl_loc + offset;
+                t_pl_loc modified_min_pl_loc = min_pl_loc - offset;
 
                 t_pl_loc max_pl_loc(reg_rect.xmax(), reg_rect.ymax(), block_regions[i].get_sub_tile());
 
-                t_pl_loc modified_max_pl_loc = max_pl_loc + offset;
+                t_pl_loc modified_max_pl_loc = max_pl_loc - offset;
 
                 modified_reg.set_region_rect(modified_min_pl_loc.x, modified_min_pl_loc.y, modified_max_pl_loc.x, modified_max_pl_loc.y);
                 //check that subtile is not an invalid value before changing, otherwise it just stays -1
-                if (block_regions[i].get_sub_tile() != -1) {
+                if (block_regions[i].get_sub_tile() != NO_SUBTILE) {
                     modified_reg.set_sub_tile(modified_min_pl_loc.sub_tile);
                 }
 
                 modified_pr.add_to_part_region(modified_reg);
             }
 
             if (num_constrained_members == 1) {
-                macro_pr = modified_pr;
+                macro_head_pr = modified_pr;
             } else {
-                macro_pr = intersection(macro_pr, modified_pr);
+                macro_head_pr = intersection(macro_head_pr, modified_pr);
             }
         }
     }
 
+    //intersect to ensure the head pr does not go outside of grid dimensions
+    macro_head_pr = intersection(macro_head_pr, grid_pr);
+
+    //if the intersection is empty, no way to place macro members together, give an error
+    if (macro_head_pr.empty()) {
+        VPR_ERROR(VPR_ERROR_PLACE, " \n Feasible floorplanning constraints could not be calculated for the placement macro.\n");
+    }
+
+    return macro_head_pr;
+}
+
+PartitionRegion update_macro_member_pr(PartitionRegion& head_pr, const t_pl_offset& offset, const PartitionRegion& grid_pr) {
+    std::vector<Region> block_regions = head_pr.get_partition_region();
+    PartitionRegion macro_pr;
+
+    for (unsigned int i = 0; i < block_regions.size(); i++) {
+        Region modified_reg;
+
+        vtr::Rect<int> reg_rect = block_regions[i].get_region_rect();
+
+        t_pl_loc min_pl_loc(reg_rect.xmin(), reg_rect.ymin(), block_regions[i].get_sub_tile());
```

Comment:
See if you can refactor this by creating an offset helper function for vtr::Rect

Reply:
Instead of adding a offset helper function for vtr::Rect, I directly added or subtract the offset to the rectangle values as needed. By doing this, the creation of the different locs was avoided, and applying the offset to the rectangle was reduced to one line.
I chose to do this instead because otherwise I would have needed to add two offset helper functions to vtr::Rect (add and subtract) and also added a mutator to the Region class to set the rectangle by passing in a vtr::Rect instead of four individual values.

Reply:
OK, if you prefer this style you can leave it.  It is explicit.